rm-datasource/src/main/java/io/seata/rm/datasource/exec/InsertExecutor.java

```diff
@@ -95,25 +95,50 @@ protected boolean containsPK() {
         List<Object> pkValues = null;
         if (statementProxy instanceof PreparedStatementProxy) {
             PreparedStatementProxy preparedStatementProxy = (PreparedStatementProxy) statementProxy;
-            ArrayList<Object>[] paramters = preparedStatementProxy.getParameters();
             int insertColumnsSize = insertColumns.size();
-            int cycleNums = paramters.length / insertColumnsSize;
-            List<Integer> pkIndexs = new ArrayList<>(cycleNums);
-            int firstPkIndex = 0;
-            for (int paramIdx = 0; paramIdx < insertColumns.size(); paramIdx++) {
+            int pkIndex = 0;
+            for (int paramIdx = 0; paramIdx < insertColumnsSize; paramIdx++) {
                 if (insertColumns.get(paramIdx).equalsIgnoreCase(pk)) {
-                    firstPkIndex = paramIdx;
+                    pkIndex = paramIdx;
                     break;
                 }
             }
-            for (int i = 0; i < cycleNums; i++) {
-                pkIndexs.add(insertColumnsSize * i + firstPkIndex);
-            }
-            if (pkIndexs.size() == 1) {
-                //adapter test case
-                pkValues = preparedStatementProxy.getParamsByIndex(pkIndexs.get(0));
-            } else {
-                pkValues = pkIndexs.stream().map(pkIndex -> paramters[pkIndex].get(0)).collect(Collectors.toList());
+
+            List<List<Object>> insertRows = recognizer.getInsertRows();
+            if (insertRows != null && !insertRows.isEmpty()) {
+                int rowSize = insertRows.size();
+                boolean hasNull = false;
+                boolean hasPlaceholder = false;
+                for (List<Object> row : insertRows) {
+                    Object pkValue = row.get(pkIndex);
+                    if (pkValue instanceof Null) {
+                        hasNull = true;
+                        continue;
+                    }
+                    if ("?".equals(pkValue)) {
+                        hasPlaceholder = true;
+                        continue;
+                    }
+                }
+                if (hasNull && hasPlaceholder) {
```

Comment:
Here we don't know pkValues yet, it is possible that the pkValue of the placeholder is also Null.

Reply:
fix done